provider/lpwindow/submit_task.go

```diff
@@ -0,0 +1,272 @@
+package lpwindow
+
+import (
+	"bytes"
+	"context"
+	"github.com/filecoin-project/go-state-types/crypto"
+	"github.com/filecoin-project/lotus/storage/wdpost"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/go-state-types/builtin"
+	"github.com/filecoin-project/go-state-types/builtin/v9/miner"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/lib/harmony/harmonydb"
+	"github.com/filecoin-project/lotus/lib/harmony/harmonytask"
+	"github.com/filecoin-project/lotus/lib/harmony/resources"
+	"github.com/filecoin-project/lotus/lib/promise"
+	"github.com/filecoin-project/lotus/provider/chainsched"
+	"github.com/filecoin-project/lotus/provider/lpmessage"
+	"github.com/filecoin-project/lotus/storage/ctladdr"
+)
+
+type WdPoStSubmitTaskApi interface {
+	ChainHead(context.Context) (*types.TipSet, error)
+
+	WalletBalance(context.Context, address.Address) (types.BigInt, error)
+	WalletHas(context.Context, address.Address) (bool, error)
+
+	StateAccountKey(context.Context, address.Address, types.TipSetKey) (address.Address, error)
+	StateLookupID(context.Context, address.Address, types.TipSetKey) (address.Address, error)
+	StateMinerInfo(context.Context, address.Address, types.TipSetKey) (api.MinerInfo, error)
+	StateGetRandomnessFromTickets(ctx context.Context, personalization crypto.DomainSeparationTag, randEpoch abi.ChainEpoch, entropy []byte, tsk types.TipSetKey) (abi.Randomness, error)
+
+	GasEstimateFeeCap(context.Context, *types.Message, int64, types.TipSetKey) (types.BigInt, error)
+	GasEstimateGasPremium(_ context.Context, nblocksincl uint64, sender address.Address, gaslimit int64, tsk types.TipSetKey) (types.BigInt, error)
+}
+
+type WdPostSubmitTask struct {
+	sender *lpmessage.Sender
+	db     *harmonydb.DB
+	api    WdPoStSubmitTaskApi
+
+	maxWindowPoStGasFee types.FIL
+	as                  *ctladdr.AddressSelector
+
+	submitPoStTF promise.Promise[harmonytask.AddTaskFunc]
+}
+
+func NewWdPostSubmitTask(pcs *chainsched.ProviderChainSched, send *lpmessage.Sender, db *harmonydb.DB, api WdPoStSubmitTaskApi, maxWindowPoStGasFee types.FIL, as *ctladdr.AddressSelector) (*WdPostSubmitTask, error) {
+	res := &WdPostSubmitTask{
+		sender: send,
+		db:     db,
+		api:    api,
+
+		maxWindowPoStGasFee: maxWindowPoStGasFee,
+		as:                  as,
+	}
+
+	if err := pcs.AddHandler(res.processHeadChange); err != nil {
```

Comment:
What happens for these case when the db entry is added after its submit epoch has started (potentially common for WinningPoSt)
   -- a "check_now()" type of function could help here.